text_generation/causal_lm/cpp/prompt_lookup_decoding_lm.cpp

```diff
@@ -0,0 +1,299 @@
+}  // namespace
+
+int main(int argc, char* argv[]) try {
+    if (argc != 3) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> '<PROMPT>'");
+    }
+
+    // tokenizer model
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+
+    const std::string model_dir = std::string{argv[1]};
+
+    auto tokenizer_model = core.read_model(model_dir + "/openvino_tokenizer.xml");
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer = core.compile_model(tokenizer_model, "CPU").create_infer_request();
+    auto [input_ids, attention_mask] = tokenize(tokenizer, argv[2]);
+
+    std::vector<int64_t> full_input_ids{input_ids.data<int64_t>(), input_ids.data<int64_t>() + input_ids.get_size()};
+
+    ov::InferRequest detokenizer =
+        core.compile_model(model_dir + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    TextStreamer text_streamer{std::move(detokenizer)};
+
+    ov::InferRequest model = core.compile_model(model_dir + "/openvino_model.xml", "CPU").create_infer_request();
+
+    model.set_tensor("input_ids", input_ids);
+    model.set_tensor("attention_mask", attention_mask);
+
+    ov::Tensor position_ids = model.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    uint64_t seq_len = input_ids.get_shape()[1];
+
+    // set beam_idx for stateful model: no beam search is used and BATCH_SIZE = 1
+    model.get_tensor("beam_idx").set_shape({BATCH_SIZE});
+    model.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+
+    // To coollect kv-cache for the <PROMPT> and to get the next token run the very first infer request
+    model.infer();
+
+    // logits shape is [BATCH_SIZE, seq_len, vocab_size]
+    auto logits = model.get_tensor("logits");
+    size_t vocab_size = logits.get_shape().back();
+    auto data_logits = logits.data<float>() + (seq_len - 1) * vocab_size;
+    int64_t out_token = std::max_element(data_logits, data_logits + vocab_size) - data_logits;
+
+    full_input_ids.push_back(out_token);
+
+    auto first_token = out_token;
+    text_streamer.put(out_token);
+
+    const int64_t EOS_TOKEN = get_eos_token(tokenizer_model);
+
+    // Prompt lookup decoding is a speculative decoding technic where the draft model replaced
+    // with string matching in the prompt to generate candidate token sequences.
+    int max_sequence_length = 100;
+    PromptLookupCandidateGenerator candidateGenerator{3, 5};
+
+    while (out_token != EOS_TOKEN && seq_len < max_sequence_length) {
+        auto candidates = candidateGenerator.generate_candidates(full_input_ids);
+        size_t candidates_size = candidates.size();
+
+        // candidates_size + 1 tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, candidates_size + 1});
```

Comment:
why in typical speculative decoding we have K tokens https://github.com/openvinotoolkit/openvino.genai/blob/master/text_generation/causal_lm/cpp/speculative_decoding_lm.cpp#L226 ? Looks unaligned ..

Reply:
It seems we throwing away last draft token in speculative decoding:
https://github.com/openvinotoolkit/openvino.genai/blob/master/text_generation/causal_lm/cpp/speculative_decoding_lm.cpp#L229

I think it should be K + 1 for speculative sample as well.

Reply:
ok, could you please discuss it with Pavel and update that sample?

Reply:
Sure, will do